File: web/lib/types/insights.ts
/**
 * Manuel TypeScript tipleri — `/api/v2/teacher/insights/*` için.
 *
 * Pydantic şemalarıyla (`app/routes/api_v2/schemas/insights.py`) birebir aynı.
 */

// =============================================================================
// Fleet overview
// =============================================================================

export interface TopPatternItem {
  book_id: number;
  book_name: string;
  section_id: number;
  section_label: string;
  subject_name: string;
  count: number;
  students: number;
}

export interface StudentMaturityItem {
  student_id: number;
  full_name: string;
  weeks_observed: number;
  days_observed: number;
  maturity_value: number;
  maturity_text: string;
  accepted_count: number;
  rejected_count: number;
  acceptance_rate: number | null;
}

export interface WeekBucketItem {
  start: string; // YYYY-MM-DD
  accepted: number;
  rejected: number;
}

export interface HealthBadge {
  key: string;
  label: string;
  color: string;
}

export interface FleetInsightsResponse {
  teacher_id: number;
  today: string;
  students: StudentMaturityItem[];
  fleet_total_accepted: number;
  fleet_total_rejected: number;
  fleet_acceptance_rate: number | null;
  avg_maturity: number;
  students_with_data: number;
  top_accepted: TopPatternItem[];
  top_rejected: TopPatternItem[];
  weekly_trend: WeekBucketItem[];
  last_activity_at: string | null;
  health_overall: HealthBadge;
  health_activity: HealthBadge;
}

// =============================================================================
// Diagnostics
// =============================================================================

export interface DiagnosticsPatternRow {
  dow: number;
  dow_label: string;
  book_id: number;
  book_name: string;
  subject_name: string;
  section_id: number;
  section_label: string;
  topic_name: string | null;
  freq: number;
  typical_count: number;
  samples: number[];
}

export interface DiagnosticsVolumeRow {
  dow: number;
  dow_label: string;
  task_count: number;
  subject_count: number;
}

export interface DiagnosticsRejectRow {
  dow_label: string;
  book_id: number;
  book_name: string;
  subject_name: string;
  section_id: number;
  section_label: string;
  weight: number;
  count: number;
  blocked: boolean;
}

export interface DiagnosticsStudentRef {
  id: number;
  full_name: string;
}

export interface StudentDiagnosticsResponse {
  student: DiagnosticsStudentRef;
  today: string;
  weeks_observed: number;
  days_observed: number;
  maturity_value: number;
  maturity_label: string;
  maturity_pct: number;
  maturity_base: number;
  maturity_floor_applied: boolean;
  maturity_weeks_constant: number;
  maturity_min_floor: number;
  reject_decay_days: number;
  reject_strong_count: number;
  reject_score_penalty: number;
  pattern_rows: DiagnosticsPatternRow[];
  volume_rows: DiagnosticsVolumeRow[];
  reject_rows: DiagnosticsRejectRow[];
  total_accepted: number;
  total_rejected: number;
}

// =============================================================================
// Suggestion panel
// =============================================================================

export interface SuggestionItem {
  book_id: number;
  book_name: string;
  book_type: string;
  section_id: number;
  section_label: string;
  subject_id: number;
  subject_name: string;
  topic_name: string | null;
  planned_count: number;
  remaining: number;
  confidence: number;
  confidence_label: string;
  score: number;
  reasons: string[];
}

export interface SuggestionDayBundle {
  date: string;
  suggestions: SuggestionItem[];
}

export interface StudentSuggestionsPanelResponse {
  student_id: number;
  target_date: string;
  suggestions: SuggestionItem[];
  maturity_value: number;
  maturity_label: string;
  weeks_observed: number;
  days_observed: number;
  active_phase: string | null;
  track_required: boolean;
  track_missing: boolean;
  track_label: string | null;
}

export interface StudentSuggestionsAheadResponse {
  student_id: number;
  today: string;
  days: SuggestionDayBundle[];
}

// =============================================================================
// Mutation bodies + results
// =============================================================================

export interface SuggestionAcceptBody {
  date: string;
  book_id: number;
  section_id: number;
  planned_count: number;
}

export interface SuggestionRejectBody {
  date: string;
  book_id: number;
  section_id: number;
}

export interface SuggestionAcceptItem {
  book_id: number;
  section_id: number;
  planned_count: number;
}

export interface SuggestionAcceptAllBody {
  date: string;
  items: SuggestionAcceptItem[];
}

export interface SuggestionAcceptResult {
  accepted: boolean;
  task_id: number;
  date: string;
}

export interface SuggestionAcceptAllResult {
  created_count: number;
  errors: string[];
}

export interface SuggestionRejectResult {
  rejected: boolean;
}

File: app/routes/api_v2/schemas/insights.py
"""API v2 — AI İçgörü, Tanılama ve Öneri şemaları (Dalga 3 Paket 9).

Kapsam:
  - Filo (fleet) içgörüleri: olgunluk haritası + sağlık ışıkları + trend
  - Öğrenci tanılaması: maturity formül parçaları + pattern/volume/reject
  - 7 günlük öneri panel + accept/reject/accept-all

Referans Jinja (dokunulmaz):
  - app/routes/teacher_ai_insights.py
  - app/routes/teacher_diagnostics.py
  - app/routes/teacher_suggestions.py

Servisler (değişmez):
  - app.services.ai_insights.build_fleet_insights
  - app.services.suggestions.{build_student_model, suggest_for_date, ...}
"""
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel


# =============================================================================
# Fleet insights — /teacher/insights/overview
# =============================================================================


HealthOverallLiteral = str  # 'no_students'|'no_data'|'warming_up'|'early'|'growing'|'mature'
HealthActivityLiteral = str  # 'never'|'today'|'recent'|'stale'|'cold'


class TopPatternItem(BaseModel):
    book_id: int
    book_name: str
    section_id: int
    section_label: str
    subject_name: str
    count: int
    students: int


class StudentMaturityItem(BaseModel):
    student_id: int
    full_name: str
    weeks_observed: int
    days_observed: int
    maturity_value: float            # 0..1
    maturity_text: str
    accepted_count: int
    rejected_count: int
    acceptance_rate: float | None    # None = yetersiz veri


class WeekBucketItem(BaseModel):
    start: date
    accepted: int
    rejected: int


class HealthBadge(BaseModel):
    key: str                         # 'no_data', 'mature', vs.
    label: str                       # Türkçe etiket
    color: str                       # CSS hex (#0ea5e9 vb.)


class FleetInsightsResponse(BaseModel):
    teacher_id: int
    today: date
    students: list[StudentMaturityItem]
    fleet_total_accepted: int
    fleet_total_rejected: int
    fleet_acceptance_rate: float | None
    avg_maturity: float
    students_with_data: int
    top_accepted: list[TopPatternItem]
    top_rejected: list[TopPatternItem]
    weekly_trend: list[WeekBucketItem]
    last_activity_at: datetime | None
    health_overall: HealthBadge
    health_activity: HealthBadge


# =============================================================================
# Öğrenci tanılaması — /teacher/insights/students/{id}/diagnostics
# =============================================================================


class DiagnosticsPatternRow(BaseModel):
    dow: int                         # 0..6
    dow_label: str
    book_id: int
    book_name: str
    subject_name: str
    section_id: int
    section_label: str
    topic_name: str | None
    freq: int
    typical_count: int
    samples: list[int]


class DiagnosticsVolumeRow(BaseModel):
    dow: int
    dow_label: str
    task_count: int                  # tipik kalem sayısı / gün
    subject_count: int               # tipik ders çeşitliliği / gün


class DiagnosticsRejectRow(BaseModel):
    dow_label: str
    book_id: int
    book_name: str
    subject_name: str
    section_id: int
    section_label: str
    weight: float
    count: int
    blocked: bool                    # REJECT_STRONG_COUNT eşiği aşıldı mı


class DiagnosticsStudentRef(BaseModel):
    id: int
    full_name: str


class StudentDiagnosticsResponse(BaseModel):
    student: DiagnosticsStudentRef
    today: date
    weeks_observed: int
    days_observed: int
    maturity_value: float
    maturity_label: str
    maturity_pct: int
    maturity_base: float
    maturity_floor_applied: bool
    maturity_weeks_constant: int
    maturity_min_floor: float
    reject_decay_days: int
    reject_strong_count: int
    reject_score_penalty: float
    pattern_rows: list[DiagnosticsPatternRow]
    volume_rows: list[DiagnosticsVolumeRow]
    reject_rows: list[DiagnosticsRejectRow]
    total_accepted: int
    total_rejected: int


# =============================================================================
# Öneri panel — /teacher/insights/students/{id}/suggestions
# =============================================================================


class SuggestionItem(BaseModel):
    book_id: int
    book_name: str
    book_type: str
    section_id: int
    section_label: str
    subject_id: int
    subject_name: str
    topic_name: str | None
    planned_count: int
    remaining: int
    confidence: float
    confidence_label: str            # 'Zayıf'|'Orta'|'Güçlü'|'Çok güçlü'
    score: float
    reasons: list[str]


class SuggestionDayBundle(BaseModel):
    date: date
    suggestions: list[SuggestionItem]


class StudentSuggestionsPanelResponse(BaseModel):
    student_id: int
    target_date: date
    suggestions: list[SuggestionItem]
    maturity_value: float
    maturity_label: str
    weeks_observed: int
    days_observed: int
    active_phase: str | None
    track_required: bool
    track_missing: bool
    track_label: str | None


class StudentSuggestionsAheadResponse(BaseModel):
    """7 günlük (today..today+6) bundle — diagnostics sekmesinde kullanılır."""
    student_id: int
    today: date
    days: list[SuggestionDayBundle]


class SuggestionAcceptBody(BaseModel):
    date: date
    book_id: int
    section_id: int
    planned_count: int               # ≥1


class SuggestionRejectBody(BaseModel):
    date: date
    book_id: int
    section_id: int


class SuggestionAcceptItem(BaseModel):
    book_id: int
    section_id: int
    planned_count: int               # ≥1


class SuggestionAcceptAllBody(BaseModel):
    date: date
    items: list[SuggestionAcceptItem]


class SuggestionAcceptResult(BaseModel):
    accepted: bool
    task_id: int
    date: date


class SuggestionAcceptAllResult(BaseModel):
    created_count: int
    errors: list[str]                # tek tek kalemlerin reddedilme sebepleri (insancıl)


class SuggestionRejectResult(BaseModel):
    rejected: bool
